### PR TITLE
Change gcloud package to google.cloud

### LINKED ISSUE #1978: Tracking namespace change
Here's a place we can track updates to the future changing of the `gcloud-python` namespace.

This change could break things in the future.

----

Sub-projects already partially created:

- [`google-cloud`](https://pypi.python.org/pypi/google-cloud)
- [`google-cloud-core`](https://pypi/.python.org/pypi/google-cloud-core)
- [`google-cloud-datastore`](https://pypi.python.org/pypi/google-cloud-datastore)
- [`google-cloud-storage`](https://pypi.python.org/pypi/google-cloud-storage)
- [`google-cloud-pubsub`](https://pypi.python.org/pypi/google-cloud-pubsub)
- [`google-cloud-bigquery`](https://pypi.python.org/pypi/google-cloud-bigquery)
- [`google-cloud-bigtable`](https://pypi.python.org/pypi/google-cloud-bigtable)
- [`google-cloud-happybase`](https://pypi.python.org/pypi/google-cloud-happybase)
- [`google-cloud-logging`](https://pypi.python.org/pypi/google-cloud-logging)
- [`google-cloud-vision`](https://pypi.python.org/pypi/google-cloud-vision)
- [`google-cloud-speech`](https://pypi.python.org/pypi/google-cloud-speech)
- [`google-cloud-language`](https://pypi.python.org/pypi/google-cloud-language) (and the alias [`google-cloud-natural-language`](https://pypi.python.org/pypi/google-cloud-natural-language))

### PR DESCRIPTION
**Pull request by dhermes**
Also use namespace packages throughout.

The commit messages will be quite useful, as almost all of them contain bash scripts used to generated the commit.

Fixes #1978 

/cc @jonparrott 

----

Currently the PR has problems (e.g. imports don't work for `tox`), but I wanted to get it out there as a start since it is such a monster.

### CODE REVIEW

File: google/__init__.py [hunk 1]
```diff
@@ -11,3 +11,10 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+try:
+    import pkg_resources
+    pkg_resources.declare_namespace(__name__)
+except ImportError:
+    import pkgutil
+    __path___ = pkgutil.extend_path(__path__, __name__)
```
Comment (jonparrott) [inline google/__init__.py]: One too many `_` at the end of `__path__` - in all files.

Comment (dhermes) [inline google/__init__.py]: Lulz

**Comment by daspecster** [comment]
@dhermes, I'm sure this will break json-docs as well.
I can start playing with that from this branch though.

**Comment by dhermes** [comment]
@daspecster Definitely. I was planning on asking for your help at some point. Trying to get the basic `tox` envs working first.

**Comment by dhermes** [comment]
@tseaver Can you hop on this ASAP?

@omaray @jgeewax The short-term fix for the `json-docs` build failure was to remove it from `.travis.yml`. As discussed earlier, I'll follow up with a script to add the readthedocs.org themed Sphinx docs into `gh-pages` and then we can go from there.

File: .travis.yml [hunk 2]
```diff
@@ -25,7 +24,7 @@ deploy:
     secure: keSFPLzVj/yAAJ8kw4m+U5kpIiGqubpDqhcHPlGAvMohFMIO4D8ffgCqJdbhQshTtb/wiS5ajK1KNpEzq4BKD/5oRXuj/sIRTMENzX/azUXkT/bADLEGocyfE4/NAGQ9mcDOmHntjt7UUhYN6H87rgex22er4V1vD5tGf5Kc1HU=
   on:
     tags: true
-    repo: GoogleCloudPlatform/gcloud-python
+    repo: GoogleCloudPlatform/google-cloud-python
     # until this is fixed: https://github.com/travis-ci/travis-ci/issues/1675
```
Comment (tseaver) [inline .travis.yml]: Is this going to break when we merge?  I'm guessing we will be renaming the repository soonish afterwards.

**Comment by tseaver** [comment]
@dhermes are the scripts you are using to generate those commits available, maybe as a gist?  The changeset is too big to review (or even show in the Github UI).

**Comment by dhermes** [comment]
@tseaver If a script was used to generate the commit, it's in the commit message. https://github.com/GoogleCloudPlatform/gcloud-python/pull/2223/commits

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -223,13 +223,13 @@ Running System Tests
    $ gcloud beta emulators datastore env-init
    export DATASTORE_LOCAL_HOST=localhost:8417
    export DATASTORE_HOST=http://localhost:8417
-   export DATASTORE_DATASET=gcloud-settings-app-id
-   export DATASTORE_PROJECT_ID=gcloud-settings-app-id
+   export DATASTORE_DATASET=google-cloud-settings-app-id
+   export DATASTORE_PROJECT_ID=google-cloud-settings-app-id
```
Comment (tseaver) [inline CONTRIBUTING.rst]: Are we sure that the output from `env-init` should be changed?

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -326,7 +326,7 @@ instead of
 ``https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/CONTRIBUTING.rst``)
 may cause problems creating links or rendering the description.
 
-.. _description on PyPI: https://pypi.python.org/pypi/gcloud
+.. _description on PyPI: https://pypi.python.org/pypi/google-cloud
```
Comment (tseaver) [inline CONTRIBUTING.rst]: We have this project reserved already?

File: docs/Makefile [hunk 5]
```diff
@@ -77,17 +77,17 @@ qthelp:
...
 
 devhelp:
 	$(SPHINXBUILD) -b devhelp $(ALLSPHINXOPTS) $(BUILDDIR)/devhelp
 	@echo
 	@echo "Build finished."
 	@echo "To view the help file:"
-	@echo "# mkdir -p $$HOME/.local/share/devhelp/gclouddatastore"
-	@echo "# ln -s $(BUILDDIR)/devhelp $$HOME/.local/share/devhelp/gclouddatastore"
+	@echo "# mkdir -p $$HOME/.local/share/devhelp/google-cloud"
+	@echo "# ln -s $(BUILDDIR)/devhelp $$HOME/.local/share/devhelp/google-cloud"
```
Comment (tseaver) [inline docs/Makefile]: Are we sure that the `datastore` portion should be dropped?

Comment (dhermes) [inline docs/Makefile]: We never use the `Makefile`. That content was in there when the original name of the project was `gcloud-datastore`.

File: CONTRIBUTING.rst (hunk 3 above)
Comment (dhermes) [inline CONTRIBUTING.rst]: Those are just placeholders. For me locally the value is `bossylobster` (my default project)

File: CONTRIBUTING.rst (hunk 4 above)
Comment (dhermes) [inline CONTRIBUTING.rst]: We do

File: google/cloud/_helpers.py [hunk 6]
```diff
@@ -42,8 +42,10 @@
 from six.moves.http_client import HTTPConnection
 from six.moves import configparser
 
+# pylint: disable=ungrouped-imports
 from google.cloud.environment_vars import PROJECT
 from google.cloud.environment_vars import CREDENTIALS
+# pylint: enable=ungrouped-imports
```
Comment (tseaver) [inline google/cloud/_helpers.py]: Ugh -- if we have those everywhere, we should maybe just disable the check altogether.

File: google/cloud/_helpers.py [hunk 7]
```diff
@@ -64,10 +66,10 @@
     _USER_ROOT = os.path.expanduser('~')
 except ImportError:  # pragma: NO COVER
     _USER_ROOT = None
-_GOOGLE_CLOUD_CONFIG_FILE = os.path.join(
+_GCLOUD_CONFIG_FILE = os.path.join(
     'gcloud', 'configurations', 'config_default')
-_GOOGLE_CLOUD_CONFIG_SECTION = 'core'
-_GOOGLE_CLOUD_CONFIG_KEY = 'project'
+_GCLOUD_CONFIG_SECTION = 'core'
+_GCLOUD_CONFIG_KEY = 'project'
```
Comment (tseaver) [inline google/cloud/_helpers.py]: Putting these back after a greedy script run, I assume?

File: google/cloud/_helpers.py (hunk 6 above)
Comment (dhermes) [inline google/cloud/_helpers.py]: Sort of yeah, sort of no. I don't even know what the right way to feel about it is. Though I think we should punt on the discussion outside of this PR.

**Comment by dhermes** [comment]
@tseaver Renames that are also a concern:

```
gcloud-python-expenses-demo
google-cloud-python.readthedocs.org
```

File: google/cloud/_helpers.py (hunk 7 above)
Comment (dhermes) [inline google/cloud/_helpers.py]: Yeah. The script was for env. vars. but these literally refer to the `gcloud` CLI

File: google/cloud/datastore/batch.py [hunk 8]
```diff
@@ -290,7 +290,7 @@ def _assign_entity_to_pb(entity_pb, entity):
 
     Helper method for ``Batch.put``.
 
-    :type entity_pb: :class:`google.cloud.datastore._generated.entity_pb2.Entity`
+    :type entity_pb: :class:`.datastore._generated.entity_pb2.Entity`
```
Comment (tseaver) [inline google/cloud/datastore/batch.py]: This doesn't look right:  it should be fully-qualified, or else `._generated....`.

File: google/cloud/datastore/client.py [hunk 9]
```diff
@@ -77,7 +78,7 @@ def _extended_lookup(connection, project, key_pbs,
     :type project: string
     :param project: The project to make the request for.
 
-    :type key_pbs: list of :class:`google.cloud.datastore._generated.entity_pb2.Key`
+    :type key_pbs: list of :class:`.datastore._generated.entity_pb2.Key`
```
Comment (tseaver) [inline google/cloud/datastore/client.py]: Ditto here.

File: google/cloud/_helpers.py (hunk 6 above)
Comment (dhermes) [inline google/cloud/_helpers.py]: Filed #2253

File: google/cloud/datastore/batch.py (hunk 8 above)
Comment (dhermes) [inline google/cloud/datastore/batch.py]: The `.` just says "start here" and find the only match. It's not a relative path.

Also the `._generated` stuff won't show up any way so it's just good habits to link to, with no tangible result (other than me staying under 80 chars here).

File: google/cloud/datastore/client.py (hunk 9 above)
Comment (dhermes) [inline google/cloud/datastore/client.py]: See above. The `.` isn't what it originally seems (I also thought it was a relative path).

File: google/cloud/streaming/exceptions.py [hunk 10]
```diff
@@ -59,7 +59,7 @@ def status_code(self):
     def from_response(cls, http_response):
         """Factory:  construct an exception from a response.
 
-        :type http_response: :class:`google.cloud.streaming.http_wrapper.Response`
+        :type http_response: :class:`~.streaming.http_wrapper.Response`
```
Comment (tseaver) [inline google/cloud/streaming/exceptions.py]: Fully-qualified again, or else `.http_wrapper.Response`.

File: google/cloud/streaming/http_wrapper.py [hunk 11]
```diff
@@ -272,10 +272,10 @@ def _check_response(response):
     :type response: :class:`Response`
     :param response: the response to validate
 
-    :raises: :exc:`google.cloud.streaming.exceptions.RequestError` if response is
-             None, :exc:`google.cloud.streaming.exceptions.BadStatusCodeError` if
+    :raises: :exc:`google.cloud.streaming.exceptions.RequestError` if response
+             is None, :exc:`~.streaming.exceptions.BadStatusCodeError` if
```
Comment (tseaver) [inline google/cloud/streaming/http_wrapper.py]: And all through here.

File: google/cloud/datastore/batch.py (hunk 8 above)
Comment (tseaver) [inline google/cloud/datastore/batch.py]: I'd say if we are going to use it, we should go whole hawg and make it look like a "normal" relative path:  you and I are not going to be the only Python programmers who trip over it.

**Comment by tseaver** [comment]
@dhermes Nope, I think we have to trust our tests here.

**Comment by dhermes** [comment]
@tseaver merging now to stop blocking other PRs and to avoid having a long Travis wait time. Will send the docstring fixes in a new PR ASAP. Also going to rename the repo immediately after merging

**Comment by dhermes** [comment]
FYI AppVeyor seems happy: https://ci.appveyor.com/project/dhermes/gcloud-python/build/1.0.631.master

**Comment by tseaver** [comment]
@dhermes System tests for pubsub and logging are failing.

**Comment by tseaver** [comment]
I'm working on a fix for them.

**Comment by dhermes** [comment]
Just filed #2259 for the overall fix. Something more specific busted for logging and pubsub?

**Comment by tseaver** [comment]
Namespace packages break the client construction for pubsub and logging.

**Pull request merged**